athena-elasticsearch/src/main/java/com/amazonaws/athena/connectors/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -147,38 +151,54 @@ protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recor
                 GeneratedRowWriter rowWriter = createFieldExtractors(recordsRequest);
 
                 // Create a new search-source injected with the projection, predicate, and the pagination batch size.
-                SearchSourceBuilder searchSource = new SearchSourceBuilder().size(QUERY_BATCH_SIZE)
+                SearchSourceBuilder searchSource = new SearchSourceBuilder()
+                        .size(QUERY_BATCH_SIZE)
                         .timeout(new TimeValue(queryTimeout, TimeUnit.SECONDS))
                         .fetchSource(ElasticsearchQueryUtils.getProjection(recordsRequest.getSchema()))
                         .query(ElasticsearchQueryUtils.getQuery(recordsRequest.getConstraints().getSummary()));
+
                 // Create a new search-request for the specified index.
-                SearchRequest searchRequest = new SearchRequest(index).preference(shard);
-                int hitsNum;
-                int currPosition = 0;
-                do {
-                    // Process the search request injecting the search-source, and setting the from position
-                    // used for pagination of results.
-                    SearchResponse searchResponse = client
-                            .getDocuments(searchRequest.source(searchSource.from(currPosition)));
-
-                    // Throw on query timeout.
+                SearchRequest searchRequest = new SearchRequest(index)
+                        .preference(shard)
+                        .scroll(TimeValue.timeValueSeconds(60))
+                        .source(searchSource.from(0));
+
+                //init scroll
+                Scroll scroll = new Scroll(TimeValue.timeValueSeconds(60));
+
+                //Read the returned scroll id, which points to the search context that’s being kept alive and will be needed in the following search scroll call
+                SearchResponse searchResponse = client.search(searchRequest, RequestOptions.DEFAULT);
+                String scrollId = searchResponse.getScrollId();
+                SearchHit[] searchHits = searchResponse.getHits().getHits();
+                Iterator<SearchHit> iterator = searchResponse.getHits().iterator();
+
+                while (searchHits != null && searchHits.length > 0 && queryStatusChecker.isQueryRunning()) {
+                    Iterator<SearchHit> finalIterator = iterator;
```

Comment:
Is this variable necessary? It looks like you can just use `iterator` as is without assigning it to `finalIterator` right?

Reply:
Also I don't think you need to even initialize `iterator` outside of the while loop. You can just do it inside where you're assigning to `finalIterator`, but instead you can just do:

`Iterator<SearchHit> iterator = searchResponse.getHits().iterator();`
on line 176 and then get rid of line 196 where you assign to it again.

Reply:
Same comment about `scrollId`, you can just move line 171 into the while loop body and then get rid of line 194.